File: worldtree/environment.py
"""
This class represents a game environment.

Created on Jun 3, 2012

@author: dscotton@gmail.com (David Scotton)
"""

import os
import sys

import pygame

from characters import enemies
from characters import powerup
from game_constants import *
import map_data
import tile

MAPS_PATH = os.path.join('media', 'maps')

# TODO: Add images here as the second argument.
EMPTY_TILE = tile.Tile()

# Map of map codes to enemy class.
ENEMIES = {
  1: enemies.Badger
}

# Map of map codes to items.  These share the same number space as the enemies, so there must
# not be any overlap.  
ITEMS = {
  129: powerup.HealthBoost
}

class Environment(object):
  """A game environment.
  
  Attributes:
    grid: A two-dimensional array of map tiles.
    surface: pygame.Surface containing the appearance of the visible part of the environment.
    screen_offset: [x, y] pixel offset of the upper-right corner of the current visible area
      from the upper-right corner of the whole map.  Must be a mutable object to support scrolling.
    height: The height of the map in number of tiles.
    width: The width of the map in number of tiles.
    dirty: boolean that's True if the map needs to be redrawn.  Is not effected by player or
      enemy movement since those aren't drawn as part of the map - only by scrolling or actually
      changing the map tiles.
    enemy_group: RendererUpdates object containing all enemy sprites.  Should be manipulated
      by the main engine and not by this class.
  """
  
  def __init__(self, map_name, offset=None):
    """Constructor.
    
    Args:
      map_name: str, name of the map in map_data dict.
      map_data: Dict containing all the layout information for a particular area. Will be
        generated by TileStudio using ts_config/worldtree.tsd.
    """
    self.name = map_name
    map_info = map_data.map_data[map_name]
    # This is a little convoluted because in order to address tiles as [x][y] (rather than
    # [y][x]) we need to build a list of columns rather than a list of rows.
    self.grid = []
    self.height = map_info['height']
    self.width = map_info['width']
    if offset is None:
      self.screen_offset = [0, 0]
    else:
      self.screen_offset = list(offset)
    self.surface = pygame.Surface(MAP_SIZE)
    self.dirty = True  # Whether the surface needs to be refreshed.
    self.enemy_group = pygame.sprite.RenderUpdates()
    self.item_group = pygame.sprite.RenderUpdates()
    image_cache = {}  # Only create one Surface for each image.
    for row in range(self.height):
      for col in range(self.width):
        if row == 0:
          # Extend the grid to width = col.
          self.grid.append([])
        if map_info['layout'][row][col] == 0:
          self.grid[col].append(EMPTY_TILE)
          continue
        
        image_name = '%s-%s.png' % (map_info['tileset'], map_info['layout'][row][col])
        if image_name not in image_cache:
          image_path = os.path.join(TILE_DIR, image_name)
          image = pygame.transform.scale(pygame.image.load(image_path), TILE_SIZE).convert_alpha()
          image_cache[image_name] = image
        self.grid[col].append(tile.Tile(image=image_cache[image_name],
                                        bound_byte=map_info['bounds'][row][col]))
        mapcode = map_info['mapcodes'][row][col]
        if mapcode != 0:
          if mapcode in ENEMIES:
            self.enemy_group.add(ENEMIES[mapcode](self, (col, row)))
          elif mapcode in ITEMS:
            self.item_group.add(ITEMS[mapcode](self, (col, row)))
          else:
            raise Exception("Unknown mapcode: %s" % mapcode)
    # TODO: Prevent enemies from walking into items.

  def VisibleTiles(self):
    """Returns the indexes of the currently visible tiles.

    Return tuple is ((first_column, last_column), (first_row, last_row)) for the visible area.
    """
    first_x = self.screen_offset[0] / TILE_WIDTH
    last_x = first_x + (MAP_WIDTH / TILE_WIDTH)
    first_y = self.screen_offset[1] / TILE_HEIGHT
    last_y = first_y + (MAP_HEIGHT / TILE_HEIGHT)
    return ((first_x, last_x), (first_y, last_y))

  def GetImage(self):
    """Get the pygame.Surface for the portion of the environment currently in the game window."""
    if self.dirty:
      self.surface.fill(BG_COLOR)
      # Figure out which tiles fit in the current window
      (first_x, last_x), (first_y, last_y) = self.VisibleTiles()
      x_pixel_start = self.screen_offset[0] % TILE_WIDTH
      y_pixel_start = self.screen_offset[1] % TILE_HEIGHT

      for col in range(first_x, last_x + 1):
        if col >= len(self.grid):
          # Outside the right edge of the grid - maybe handle this better?
          break
        for row in range(first_y, last_y + 1):
          if row >= len(self.grid[col]):
            # Past the bottom of the grid.
            break
          tile_x_pos = (col - first_x) * TILE_WIDTH - x_pixel_start
          tile_y_pos = (row - first_y) * TILE_HEIGHT - y_pixel_start
          try:
            self.surface.blit(self.grid[col][row].image, (tile_x_pos, tile_y_pos))
          except TypeError:
            print col, row, type(self.grid[col][row])

      self.dirty = False
    return self.surface

  def AttemptMove(self, sprite, vector):
    """Checks whether a sprite's attempted movement is legal.

    Args:
      sprite: pygame.sprite.Sprite that's trying to move.
      vector: the x, y motion vector the sprite is trying to move along.
    Returns:
      A Rect for the position the sprite ends up in based on its motion and interaction with
      the environment.
    """
    # Hack to work around the environment not beginning at the top of the screen
    hitbox = sprite.Hitbox()
    dest = hitbox.move(vector)
    new_vector = list(vector)
    # Figure out which tiles contain the old and new position.
    old_tiles = self.TilesForRect(hitbox)
    new_tiles = [t for t in self.TilesForRect(dest) if t not in old_tiles]
    # For each new tile the sprite would occupy, check whether it blocks movement in the
    # desired direction.
    for col, row in new_tiles:
      tile_rect = self.RectForTile(col, row)
      # Stop non-player sprites from moving outside the room.  Allow players to move this way,
      # And check for transitions in the main loop.
      if col < 0 or col >= self.width or row < 0 or row >= self.height:
        if sprite.IS_PLAYER:
          square = tile.Tile(solid=(False, False, False, False))
        else:
          square = tile.Tile(solid=(True, True, True, True))
      else:
        square = self.grid[col][row]

      # Handle motion in each cardinal direction separately.  Need to check three conditions:
      # that sprite was previously on a particular side of the tile, and that entry from that
      # side is forbidden, and that the movement in this direction hasn't already been stopped
      # short.
      if hitbox.bottom < tile_rect.top and square.solid_top and dest.bottom >= tile_rect.top:
        new_vector[1] = tile_rect.top - hitbox.bottom - 1
      elif (hitbox.top > tile_rect.bottom and square.solid_bottom
            and dest.top <= tile_rect.bottom):
        new_vector[1] = tile_rect.bottom - hitbox.top + 1
      if hitbox.right < tile_rect.left and square.solid_left and dest.right >= tile_rect.left:
        new_vector[0] = tile_rect.left - hitbox.right - 1
      elif (hitbox.left > tile_rect.right and square.solid_right
            and dest.left <= tile_rect.right):
        new_vector[0] = tile_rect.right - hitbox.left + 1

    new_position = sprite.rect.move(new_vector)
    return new_position
    
  def TilesForRect(self, rect):
    """Returns a set of tiles that a rect falls in.

    The rect is a position on the map, including any portions currently offscreen.
    """
    left_col = rect.left / TILE_WIDTH
    right_col = rect.right / TILE_WIDTH
    top_row = rect.top / TILE_HEIGHT
    bottom_row = rect.bottom / TILE_HEIGHT
    return [(col, row) for col in range (left_col, right_col+1)
            for row in range(top_row, bottom_row+1)]

  def RectForTile(self, col, row):
    """Return a Rect object for a particular tile, from its map column and row.
    
    This rect is relative to the origin of the map, not the screen.
    """
    left = col * TILE_WIDTH
    top = row * TILE_HEIGHT
    return pygame.Rect(left, top, TILE_WIDTH-1, TILE_HEIGHT-1)

  def TileIndexForPoint(self, x, y):
    """Return the col, row index for the tile containing map coordinate (x, y)."""
    return (x / TILE_WIDTH, y / TILE_HEIGHT)

  def ScreenCoordinateForMapPoint(self, x, y):
    """Convert map-relative (x, y) pixel coordinates into screen-relative coordinates."""
    return (x + MAP_X - self.screen_offset[0], y + MAP_Y - self.screen_offset[1])
  
  def MapCoordinateForScreenPoint(self, x, y):
    """Convert screen-relative (x, y) point into a map-relative coordinate."""
    return (x - MAP_X + self.screen_offset[0], y - MAP_Y + self.screen_offset[1])

  def IsOutsideMap(self, rect):
    """True if the center of rect has moved outside the map."""
    col, row = self.TileIndexForPoint(rect.centerx, rect.centery)
    if col < 0 or col >= self.width or row < 0 or row >= self.height:
      return True
    return False

  def IsRectSupported(self, rect):
    """Returns true if there is a solid tile directly under a rectangle.
    
    Args:
      rect: A Rect object whose position is relative to the map origin (not the screen origin).
    """
    dest = rect.move((0, 1))
    old_tiles = self.TilesForRect(rect)
    new_tiles = [tile for tile in self.TilesForRect(dest) if tile not in old_tiles]
    for col, row in new_tiles:
      if col < 0 or col >= self.width:
        continue
      if row >= self.height:
        return False
      try:
        if self.grid[col][row].solid_top:
          return True
      except IndexError:
        print col, row
        raise
    return False

  def IsTileSupported(self, col, row):
    """Returns true if there is a solid tile directly under the tile being checked."""
    return self.IsRectSupported(self.RectForTile(col, row))

  def Scroll(self, rect):
    """If necessary, scroll the map to follow the position of rect.
    
    Returns:
      (x, y) vector to apply to rect to maintain its relative position with the landscape.
    """
    # x_scroll and y_scroll are calculated with opposite signs (meaning subtracted from the
    # current screen offset) in order to correctly move characters around the map to maintain
    # relative position.
    x_scroll = 0
    y_scroll = 0
    if rect.centerx < SCROLL_MARGIN and self.screen_offset[0] > 0:
      x_scroll = min(SCROLL_MARGIN - rect.centerx, self.screen_offset[0])
      self.screen_offset[0] = self.screen_offset[0] - x_scroll
      self.dirty = True
    elif (rect.centerx > MAP_WIDTH - SCROLL_MARGIN
          and self.screen_offset[0] + MAP_WIDTH < self.width * TILE_WIDTH):
      x_scroll = max(MAP_WIDTH - SCROLL_MARGIN - rect.centerx,
                     self.screen_offset[0] - self.width * TILE_WIDTH - MAP_WIDTH)
      self.screen_offset[0] = self.screen_offset[0] - x_scroll
      self.dirty = True
      
    if rect.centery < SCROLL_MARGIN + MAP_Y and self.screen_offset[1] > 0:
      y_scroll = min(SCROLL_MARGIN + MAP_Y - rect.centery, self.screen_offset[1])
      self.screen_offset[1] = self.screen_offset[1] - y_scroll
      self.dirty = True
    elif (rect.centery > MAP_HEIGHT + MAP_Y - SCROLL_MARGIN
          and self.screen_offset[1] + MAP_HEIGHT < self.height * TILE_HEIGHT):
      y_scroll = max(MAP_HEIGHT + MAP_Y - SCROLL_MARGIN - rect.centery,
                     self.screen_offset[1] + MAP_Y - self.height * TILE_HEIGHT - MAP_HEIGHT)
      self.screen_offset[1] = self.screen_offset[1] - y_scroll
      self.dirty = True

    scroll_vector = (x_scroll, y_scroll)
    # Must move all enemies to account for the shifted window.
    for enemy in self.enemy_group:
      enemy.rect = enemy.rect.move(scroll_vector)
    for item in self.item_group:
      item.rect = item.rect.move(scroll_vector)
    return scroll_vector

File: worldtree-raylib/src/Environment.cs
// src/Environment.cs
using Raylib_cs;
using System.Numerics;

namespace WorldTree;

/// <summary>
/// A game environment (one room/map).
/// Corresponds to worldtree/environment.py.
/// Entity positions are in world (map) coordinates.
/// </summary>
public class Environment
{
    // Static maps loaded once at startup
    public static Dictionary<int, Dictionary<string, MapInfo>> Regions = new();
    // {region: {room: song_filename}}
    public static Dictionary<int, Dictionary<string, string>> SongsByRoom = new();
    // {region: {room: Color}}
    public static Dictionary<int, Dictionary<string, Color>> BgColorsByRoom = new();
    // Transitions
    public static Dictionary<int, Dictionary<string, Dictionary<TransitionDirection, List<TransitionInfo>>>> AllTransitions = new();

    public string Name { get; }
    public int Region { get; }
    public int Width { get; }   // in tiles
    public int Height { get; }  // in tiles
    public Color BgColor { get; }

    // Tile grid indexed as [col][row]
    private Tile[][] _grid;

    // Entity groups (world coordinates)
    public List<Character> EnemyGroup { get; } = new();
    public List<DyingAnimation> DyingAnimationGroup { get; } = new();
    public List<Powerup> ItemGroup { get; } = new();
    public List<Projectile> HeroProjectileGroup { get; } = new();
    public List<Projectile> EnemyProjectileGroup { get; } = new();

    /// <summary>
    /// Internal constructor for testing. Bypasses asset loading.
    /// </summary>
    internal Environment(int width, int height, Tile[][] grid)
    {
        Name = "TestMap";
        Region = 1;
        Width = width;
        Height = height;
        BgColor = Color.Black;
        _grid = grid;
    }

    public Environment(string mapName, int region)
    {
        Name = mapName;
        Region = region;
        var mapInfo = Regions[region][mapName];
        Width = mapInfo.Width;
        Height = mapInfo.Height;
        BgColor = BgColorsByRoom.TryGetValue(region, out var roomColors)
                  && roomColors.TryGetValue(mapName, out var c) ? c : Color.Black;

        _grid = new Tile[Width][];
        for (int col = 0; col < Width; col++)
            _grid[col] = new Tile[Height];

        BuildTileGrid(mapInfo);
    }

    private void BuildTileGrid(MapInfo mapInfo)
    {
        var imageCache = new Dictionary<string, Texture2D>();
        var areas = new Dictionary<int, List<(int col, int row)>>();

        for (int row = 0; row < Height; row++)
        {
            for (int col = 0; col < Width; col++)
            {
                int tileId = mapInfo.Layout[row][col];
                int boundByte = mapInfo.Bounds[row][col];
                int mapCode = mapInfo.Mapcodes[row][col];

                if (tileId == 0)
                {
                    _grid[col][row] = Tile.Empty;
                }
                else
                {
                    string imageName = $"{mapInfo.Tileset}-{tileId}.png";
                    if (!imageCache.TryGetValue(imageName, out var tex))
                    {
                        tex = TextureCache.LoadTile(
                            Path.Combine(GameConstants.TileDir, imageName));
                        imageCache[imageName] = tex;
                    }
                    _grid[col][row] = Tile.WithImage(tex, (byte)boundByte);
                }

                if (mapCode != 0)
                    SpawnMapCode(mapCode, col, row, areas);
            }
        }

        CreateAreas(areas);
    }

    // Enemy/item spawn codes — mirrors ENEMIES and ITEMS dicts in environment.py
    private static readonly HashSet<int> AreaCodes = new() { 254, 255 };

    private void SpawnMapCode(int mapCode, int col, int row,
        Dictionary<int, List<(int, int)>> areas)
    {
        if (AreaCodes.Contains(mapCode))
        {
            areas.TryAdd(mapCode, new List<(int, int)>());
            areas[mapCode].Add((col, row));
            return;
        }
        var pos = (col, row);
        Character? enemy = mapCode switch
        {
            1  => new Enemies.Beaver(this, pos),
            2  => new Enemies.Dragonfly(this, pos),
            3  => new Enemies.BoomBug(this, pos),
            4  => new Enemies.Shooter(this, pos),
            5  => new Enemies.BugPipe(this, pos),
            6  => new Enemies.PipeBug(this, pos),
            7  => new Enemies.Batzor(this, pos),
            8  => new Enemies.BiterPipe(this, pos),
            9  => new Enemies.Biter(this, pos),
            10 => new Enemies.Slug(this, pos),
            11 => new Enemies.Baron(this, pos),
            _ => null
        };
        if (enemy != null) { EnemyGroup.Add(enemy); return; }

        Powerup? item = mapCode switch
        {
            129 => new Powerups.HealthBoost(this, pos),
            130 => new Powerups.DoubleJump(this, pos),
            131 => new Powerups.MoreSeeds(this, pos),
            _ => null
        };
        if (item != null) { ItemGroup.Add(item); return; }

        throw new Exception($"Unknown mapcode: {mapCode}");
    }

    private void CreateAreas(Dictionary<int, List<(int col, int row)>> areaDict)
    {
        foreach (var (mapCode, coords) in areaDict)
        {
            int i = 0;
            while (i < coords.Count)
            {
                var start = coords[i];
                int width = 1;
                while (i + width < coords.Count
                       && coords[i + width].row == start.row
                       && coords[i + width].col == start.col + width)
                    width++;
                Powerup area = mapCode switch
                {
                    254 => new Powerups.Spike(this, start, width),
                    255 => new Powerups.Lava(this, start, width),
                    _   => throw new Exception($"Unknown area code: {mapCode}")
                };
                ItemGroup.Add(area);
                i += width;
            }
        }
    }

    public Tile GetTile(int col, int row)
    {
        if (col < 0 || col >= Width || row < 0 || row >= Height)
            return Tile.Empty;
        return _grid[col][row];
    }

    // World rect for a tile (used for collision)
    public Rectangle RectForTile(int col, int row) =>
        new Rectangle(col * GameConstants.TileWidth, row * GameConstants.TileHeight,
                      GameConstants.TileWidth - 1, GameConstants.TileHeight - 1);

    public (int col, int row) TileIndexForPoint(float x, float y) =>
        ((int)MathF.Floor(x / GameConstants.TileWidth),
         (int)MathF.Floor(y / GameConstants.TileHeight));

    public bool IsOutsideMap(Rectangle rect)
    {
        var (col, row) = TileIndexForPoint(rect.CenterX(), rect.CenterY());
        return col < 0 || col >= Width || row < 0 || row >= Height;
    }

    // Camera state (replaces screen_offset in Python)
    public Vector2 ScreenOffset { get; private set; } = Vector2.Zero;

    public Camera2D MakeCamera() => new Camera2D
    {
        Offset = new Vector2(GameConstants.MapX, GameConstants.MapY),
        Target = ScreenOffset,
        Rotation = 0f,
        Zoom = 1f
    };

    /// <summary>
    /// Update the camera to follow rect (world coords). Returns updated camera.
    /// Replaces Environment.Scroll() in Python — no need to apply scroll_vector to entities
    /// because all entities are already in world coordinates.
    /// </summary>
    public Camera2D Scroll(Camera2D camera, Rectangle rect)
    {
        var offset = ScreenOffset;
        float mapPixelW = Width * GameConstants.TileWidth;
        float mapPixelH = Height * GameConstants.TileHeight;

        if (rect.CenterX() < offset.X + GameConstants.ScrollMarginX && offset.X > 0)
            offset.X = MathF.Max(0, rect.CenterX() - GameConstants.ScrollMarginX);
        else if (rect.CenterX() > offset.X + GameConstants.MapWidth - GameConstants.ScrollMarginX
                 && offset.X + GameConstants.MapWidth < mapPixelW)
            offset.X = MathF.Min(mapPixelW - GameConstants.MapWidth,
                                 rect.CenterX() - (GameConstants.MapWidth - GameConstants.ScrollMarginX));

        if (rect.CenterY() < offset.Y + GameConstants.ScrollMarginY && offset.Y > 0)
            offset.Y = MathF.Max(0, rect.CenterY() - GameConstants.ScrollMarginY);
        else if (rect.CenterY() > offset.Y + GameConstants.MapHeight - GameConstants.ScrollMarginY
                 && offset.Y + GameConstants.MapHeight < mapPixelH)
            offset.Y = MathF.Min(mapPixelH - GameConstants.MapHeight,
                                 rect.CenterY() - (GameConstants.MapHeight - GameConstants.ScrollMarginY));

        ScreenOffset = offset;
        camera.Target = offset;
        return camera;
    }

    /// <summary>Set initial camera offset (e.g. on room transition).</summary>
    public void SetScreenOffset(float x, float y) =>
        ScreenOffset = new Vector2(
            Math.Clamp(x, 0, Width * GameConstants.TileWidth - GameConstants.MapWidth),
            Math.Clamp(y, 0, Height * GameConstants.TileHeight - GameConstants.MapHeight));

    /// <summary>
    /// Draw the visible tiles. Call this inside BeginMode2D / EndMode2D.
    /// </summary>
    public void DrawTiles()
    {
        int firstCol = (int)(ScreenOffset.X / GameConstants.TileWidth);
        int lastCol = firstCol + GameConstants.MapWidth / GameConstants.TileWidth + 1;
        int firstRow = (int)(ScreenOffset.Y / GameConstants.TileHeight);
        int lastRow = firstRow + GameConstants.MapHeight / GameConstants.TileHeight + 1;

        for (int col = firstCol; col <= Math.Min(lastCol, Width - 1); col++)
            for (int row = firstRow; row <= Math.Min(lastRow, Height - 1); row++)
            {
                var tile = _grid[col][row];
                if (tile == Tile.Empty || tile.Image == null) continue;
                Raylib.DrawTexture(tile.Image.Value,
                    col * GameConstants.TileWidth,
                    row * GameConstants.TileHeight,
                    Color.White);
            }
    }

    /// <summary>
    /// Draw solid-edge indicators for all visible tiles. Call inside BeginMode2D/EndMode2D.
    /// Each solid edge is shown as a white line along that side of the tile.
    /// </summary>
    public void DrawDebugBounds()
    {
        int firstCol = (int)(ScreenOffset.X / GameConstants.TileWidth);
        int lastCol  = firstCol + GameConstants.MapWidth  / GameConstants.TileWidth + 1;
        int firstRow = (int)(ScreenOffset.Y / GameConstants.TileHeight);
        int lastRow  = firstRow + GameConstants.MapHeight / GameConstants.TileHeight + 1;

        for (int col = firstCol; col <= Math.Min(lastCol, Width - 1); col++)
        {
            for (int row = firstRow; row <= Math.Min(lastRow, Height - 1); row++)
            {
                var tile = _grid[col][row];
                if (tile.IsEmpty) continue;

                int x  = col * GameConstants.TileWidth;
                int y  = row * GameConstants.TileHeight;
                int x2 = x + GameConstants.TileWidth;
                int y2 = y + GameConstants.TileHeight;

                if (tile.SolidTop)    Raylib.DrawLine(x, y,  x2, y,  Color.White);
                if (tile.SolidBottom) Raylib.DrawLine(x, y2, x2, y2, Color.White);
                if (tile.SolidLeft)   Raylib.DrawLine(x, y,  x,  y2, Color.White);
                if (tile.SolidRight)  Raylib.DrawLine(x2, y, x2, y2, Color.White);
            }
        }
    }

    public bool IsWorldPointVisible(float x, float y) =>
        x >= ScreenOffset.X && x <= ScreenOffset.X + GameConstants.MapWidth &&
        y >= ScreenOffset.Y && y <= ScreenOffset.Y + GameConstants.MapHeight;

    /// <summary>
    /// Check which tile grid cells a world-coordinate rect intersects.
    /// </summary>
    public List<(int col, int row)> TilesForRect(Rectangle rect)
    {
        int left = (int)MathF.Floor(rect.Left() / GameConstants.TileWidth);
        int right = (int)MathF.Floor(rect.Right() / GameConstants.TileWidth);
        int top = (int)MathF.Floor(rect.Top() / GameConstants.TileHeight);
        int bot = (int)MathF.Floor(rect.Bottom() / GameConstants.TileHeight);
        var result = new List<(int, int)>();
        for (int c = left; c <= right; c++)
            for (int r = top; r <= bot; r++)
                result.Add((c, r));
        return result;
    }

    /// <summary>
    /// Attempt a move, blocking per-tile-side solidity. Returns new world rect.
    /// Resolves Y first, then X, so upward movement clears obstacles before
    /// horizontal movement is checked against them.
    /// For player (isPlayer=true), allows moving off map edges for room transitions.
    /// </summary>
    public Rectangle AttemptMove(Rectangle hitbox, (float x, float y) vector, bool isPlayer = false)
    {
        var afterY = MoveY(hitbox, vector.y, isPlayer);
        return MoveX(afterY, vector.x, isPlayer);
    }

    private Rectangle MoveY(Rectangle hitbox, float dy, bool isPlayer)
    {
        if (dy == 0f) return hitbox;
        var dest = hitbox.Move(0, dy);
        float newVY = dy;
        var oldTiles = new HashSet<(int, int)>(TilesForRect(hitbox));
        foreach (var (col, row) in TilesForRect(dest).Where(t => !oldTiles.Contains(t)))
        {
            Tile square;
            if (col < 0 || col >= Width || row < 0 || row >= Height)
                square = isPlayer ? new Tile(null, false, false, false, false)
                                  : new Tile(null, true, true, true, true);
            else
                square = _grid[col][row];

            var tileRect = RectForTile(col, row);

            if (hitbox.Bottom() < tileRect.Top() && square.SolidTop && dest.Bottom() >= tileRect.Top())
                newVY = MathF.Min(newVY, tileRect.Top() - hitbox.Bottom() - 1);
            else if (hitbox.Top() > tileRect.Bottom() && square.SolidBottom && dest.Top() <= tileRect.Bottom())
                newVY = MathF.Max(newVY, tileRect.Bottom() - hitbox.Top() + 1);
        }
        return hitbox.Move(0, newVY);
    }

    private Rectangle MoveX(Rectangle hitbox, float dx, bool isPlayer)
    {
        if (dx == 0f) return hitbox;
        var dest = hitbox.Move(dx, 0);
        float newVX = dx;
        var oldTiles = new HashSet<(int, int)>(TilesForRect(hitbox));
        foreach (var (col, row) in TilesForRect(dest).Where(t => !oldTiles.Contains(t)))
        {
            Tile square;
            if (col < 0 || col >= Width || row < 0 || row >= Height)
                square = isPlayer ? new Tile(null, false, false, false, false)
                                  : new Tile(null, true, true, true, true);
            else
                square = _grid[col][row];

            var tileRect = RectForTile(col, row);

            if (hitbox.Right() < tileRect.Left() && square.SolidLeft && dest.Right() >= tileRect.Left())
                newVX = MathF.Min(newVX, tileRect.Left() - hitbox.Right() - 1);
            else if (hitbox.Left() > tileRect.Right() && square.SolidRight && dest.Left() <= tileRect.Right())
                newVX = MathF.Max(newVX, tileRect.Right() - hitbox.Left() + 1);
        }
        return hitbox.Move(newVX, 0);
    }

    /// <summary>
    /// Simplified collision check for projectiles — off-map is illegal.
    /// </summary>
    public bool IsMoveLegal(Rectangle hitbox, (float x, float y) vector)
    {
        var dest = hitbox.Move(vector.x, vector.y);
        var oldTiles = new HashSet<(int, int)>(TilesForRect(hitbox));
        foreach (var (col, row) in TilesForRect(dest).Where(t => !oldTiles.Contains(t)))
        {
            Tile square;
            if (col < 0 || col >= Width || row < 0 || row >= Height)
                square = new Tile(null, true, true, true, true);
            else
                square = _grid[col][row];

            var tileRect = RectForTile(col, row);

            if (hitbox.Bottom() < tileRect.Top() && square.SolidTop && dest.Bottom() >= tileRect.Top())
                return false;
            if (hitbox.Top() > tileRect.Bottom() && square.SolidBottom && dest.Top() <= tileRect.Bottom())
                return false;
            if (hitbox.Right() < tileRect.Left() && square.SolidLeft && dest.Right() >= tileRect.Left())
                return false;
            if (hitbox.Left() > tileRect.Right() && square.SolidRight && dest.Left() <= tileRect.Right())
                return false;
        }
        return true;
    }

    /// <summary>
    /// Returns true if there is a solid tile in the direction of vector from rect.
    /// </summary>
    public bool IsRectSupported(Rectangle rect, (float x, float y) vector = default)
    {
        if (vector == default) vector = (0, 1);
        var dest = rect.Move(vector.x, vector.y);
	var oldTiles = new HashSet<(int, int)>(TilesForRect(rect));
        foreach (var (col, row) in TilesForRect(dest).Where(t => !oldTiles.Contains(t)))
        {
            if (col < 0 || col >= Width) continue;
            if (row < 0) return false;
            if (row >= Height) return true;
            if (_grid[col][row].SolidTop) return true;
        }
        return false;
    }

    public bool IsTileSupported(int col, int row) =>
        IsRectSupported(RectForTile(col, row));
}
